src/main/java/com/azure/cosmos/sample/ReadManyItems.java

```diff
@@ -76,178 +79,165 @@ public static void main(String[] args) {
+        // collect the ids that were generated when writing the data.
         List<String> list = new ArrayList<String>();
-        for (final String id : this.docIDs) {
-            list.add(id);
-        }
-        final ExecutorService es = Executors.newCachedThreadPool();
-        List<List<String>> lists = Lists.partition(list, NUMBER_OF_REQUESTS_PER_THREAD);
-        final long totalStartTime = System.currentTimeMillis();
-        for (List<String> splitList : lists) {
-            final Runnable task = () -> {
-                for (final String id : splitList) {
-                    try {
-                        try {
-                            //read each item in this chunk sequentially
-                            CosmosItemResponse<Item> item = container.readItem(id, new PartitionKey(id), Item.class);
-                            double requestCharge = item.getRequestCharge(); 
-                            BinaryOperator<Double> add
-                            = (u, v) -> u + v;
-                            totalRequestCharges.getAndAccumulate(requestCharge, add);
-                        
-                        } catch (CosmosException e) {
-                            System.out.println("Read Item failed with"+e);
-                        }
-                    } catch (final Exception e) {
-                        exceptionCount.incrementAndGet();
-                        System.out.println("Exception: " + e);
-                    }
-                }
-            };
-            es.execute(task);
-        }
-        es.shutdown();
-        final boolean finished = es.awaitTermination(5, TimeUnit.MINUTES);
-        final long totalEndTime = System.currentTimeMillis();
-        final long totalDuration = (totalEndTime - totalStartTime);        
-        totalReadLatency.getAndAdd(totalDuration);
-        if (finished) {
-            Thread.sleep(1000);
+        for (final JsonNode doc : docs) {
+            list.add(doc.get("id").asText());
         }
+        List<List<String>> lists = Lists.partition(list, NUMBER_OF_DOCS_PER_THREAD);
+
+        final long startTime = System.currentTimeMillis();
+        Flux.fromIterable(lists).flatMapIterable(x -> x)
+                .flatMap(id -> container.readItem(id, new PartitionKey(id), JsonNode.class))
+                .flatMap(itemResponse -> {
+                    if (itemResponse.getStatusCode() == 200) {
+                        double requestCharge = itemResponse.getRequestCharge();
+                        BinaryOperator<Double> add = (u, v) -> u + v;
+                        totalRequestCharges.getAndAccumulate(requestCharge, add);
+                        number_docs_read.getAndIncrement();
+                    } else
+                        System.out.println("WARNING insert status code {} != 200" + itemResponse.getStatusCode());
+                    return Mono.empty();
+                }).subscribe();
+        final long endTime = System.currentTimeMillis();
```

Comment:
end time will be incorrect here, since we are subscribing to the operation, but we are not sure if it will complete by the time code reaches this line.

Reply:
We need to add `onComplete()` API call or something like what you have been using above to correctly capture endTime.

Reply:
Thanks, used the while on request_count to achieve the desired effect.